doc/design/reader/low_level_reader.md

```diff
@@ -0,0 +1,71 @@
+# Paddle Low Level Data Reader
+
+This design doc will describe a way to implement C++ written data reader(feeder) to
+feed data to paddle when train or inference inorder to improve performance, and make
+best usage of CPU and GPU usage.
+
+This design can be used for both V2 API and new operator design.
+
+## Background
+
+We have sevaral issues talking about training with v2 API get bad training performance,
+like:
+- https://github.com/PaddlePaddle/Paddle/issues/3675 and
+- https://github.com/PaddlePaddle/Paddle/issues/4156 and
+- https://github.com/PaddlePaddle/Paddle/issues/3600.
+
+The current implementations for data reader can greatly simplify user defining how
+to parse and feed the data to neural network, but we need to copy data from python
+to C++ when feeding the data, and the data feeder will wait until the calculations
+of one mini-batch is done.
+
+In V1 implementations there are some considerations to improve data feeding
+performance, we can port these features to improve the reader.
+
+## Implementaion
+
+A low-level reader contains below components:
+
+1. Data parser:
+    A data parser will use `dlopen` to open user-defined parser plugin, and call
```

Comment:
With the current V2 API data reader, user can implement the C++ parser as a Python module, and call it from the user-defined data reader. Maybe we don't have to do anything on our side.

After the refactor, reader will be an OP. In this case we can have a post-process OP that allow `dlopen` any `*.so` file. It's a cool idea.

Curious what is the priority of this feature? e.g., any feature request from the user that need the custom post-processing code? (I think there are a limited set of post-processing, such as image enhancements and subtract mean, maybe we can provide the corresponding OP so users don't have to write their own version). Just my 2 cents.

Reply:
> user can implement the C++ parser as a Python module, and call it from the user-defined data reader. Maybe we don't have to do anything on our side.

Do you mean by "Python extension"? If so, maybe this will be less efficient because the data must be transferred though `python-extension.so -> python -> numpy(extension too) -> paddle C++`

> what is the priority of this feature?

Improving v2 reader performance is of high priority for users need to improve their jobs.

> any feature request from the user that need the custom post-processing code

What do you mean by "post-processing"?

Reply:
Thanks for the reply!

> Do you mean by "Python extension"? If so, maybe this will be less efficient because the data must be transferred though python-extension.so -> python -> numpy(extension too) -> paddle C++

Yes, that what I mean. This cross language inefficiency (but I am not sure if it's a bottleneck) is due to the fact that the data reader is in Python. I think if we want to eliminate this part of inefficiency, the user should use the `read data OP` after refactor.

> What do you mean by "post-processing"?

My bad, what in my mind is that we will have a `binary reader OP`, so the `MyFeeder` class (that converts bytes to Tensor) you mentioned is actually a "post-processing" module. We are eventually talking about the same thing. But I think we should classify it as a "post-processing" module, in this way the implementation only cares about the parsing and don't need to care about the reading (potentially from different filesystems).

> what is the priority of this feature?

Sorry, I meant what is the priority of allowing user to write custom feeder OP (equivalent to post-processing OP, reason mentioned above) v.s. we implement many common post-processing OP for the user to use.

I agree that "Improving v2 reader performance is of high priority for users need to improve their jobs.", but the reader performance will be solved once we have a `reader OP` and `queue OP`. I think "if we allow user to write custom *.so file and able to `dlopen` from PaddlePaddle" is more a flexibility problem than a performance problem.